akka-stream/src/main/scala/akka/stream/impl/StreamLayout.scala

```diff
@@ -550,6 +620,7 @@ private[stream] abstract class MaterializerSession(val topLevel: StreamLayout.Mo
         case mv: MaterializedValueSource[_] ⇒
           val pub = new MaterializedValuePublisher
           materializedValuePublishers ::= pub
+          materializedValues.put(mv, ())
```

Comment:
You should run the materialization benchmarks (or add a new one) to see how much impact does this have (given that the default is Keep.left almost always). If it is significantly slower ( > 5%) then I am against this.